lib/trino-aws/src/main/java/io/trino/aws/AwsCredentialsProviderFactory.java

```diff
@@ -0,0 +1,161 @@
+import static java.lang.String.format;
+import static java.util.Objects.requireNonNull;
+
+/**
+ * This factory can be used to create an {@link AWSCredentialsProvider}
+ * based on an arbitrary {@link AwsCredentialsProviderConfig configuration}.
+ * <p>
+ * The {@link AWSCredentialsProvider} returned by this factory will depend on
+ * the configuration in the following way:
+ * <ul>
+ *     <li>Static credentials embedded in the {@link AwsCredentialsProviderConfig#getAwsServiceUri()},
+ *     if present,</li>
+ *     <li>Or, a custom implementation if {@link AwsCredentialsProviderConfig#getAwsCredentialsProvider()}
+ *     is present and the class named by it implements the {@link AWSCredentialsProvider}
+ *     interface. A custom {@link CustomProviderResolver} can be provided via
+ *     {@link AwsCredentialsProviderConfig#getCustomProviderResolver()}; the default one
+ *     requires a no-argument constructor and uses plain {@link Class#forName(String)}.</li>
+ * </ul>
+ * Otherwise,
+ * <ul>
+ *     <li>If {@link AwsCredentialsProviderConfig#getAwsAccessKey()} and
+ *     {@link AwsCredentialsProviderConfig#getAwsSecretKey()} both are present, they
+ *     will be returned as static credentials. These credentials may also include
+ *     an optional {@link AwsCredentialsProviderConfig#getAwsSessionToken()}.</li>
+ *     <li>Otherwise, the default credentials provider will be returned,
+ *     which will try to obtain the credentials in the following order:
+ *     <ol>
+ *         <li>Environment variables</li>
+ *         <li>System properties</li>
+ *         <li>Web Identity token</li>
+ *         <li>Local profile ({@code ~/.aws/credentials})</li>
+ *         <li>Local EC2 instance metadata</li>
+ *     </ol>
+ *     </li>
+ *     <li>In addition, if {@link AwsCredentialsProviderConfig#getIamRole()}
+ *     is present, the returned credential provider will retrieve the credentials
+ *     having assumed the named role; it can also use the optional
+ *     {@link AwsCredentialsProviderConfig#getExternalId()} while assuming the role.
+ *     If that is the case, then {@link AwsCredentialsProviderConfig#getAwsRegionName()}
+ *     can also be provided.</li>
+ * </ul>
+ */
+public class AwsCredentialsProviderFactory
+{
+    private static final AwsRegionProviderChain REGION_PROVIDER_CHAIN = new AwsRegionProviderChain(
+            new DefaultAwsRegionProviderChain(),
+            // fallback:
+            new AwsRegionProvider()
+            {
+                @Override
+                public String getRegion()
+                {
+                    return US_EAST_1.getName();
+                }
+            });
+    private static final Pattern EMBEDDED_CREDENTIALS_PATTERN = Pattern.compile("([^:]+):(.+)");
+
+    public AWSCredentialsProvider createAwsCredentialsProvider(AwsCredentialsProviderConfig config)
+    {
+        requireNonNull(config, "config is null");
+
+        return config.getAwsServiceUri()
```

Comment:
This would be new functionality so feel free to ignore.

It would be nice to have the config options checked explicitly so that only ones which are used can be specified. For example, if you set `externalId` but not the iam role, it would make sense to fail startup with a configuration error